### PR TITLE
Detailed when

### LINKED ISSUE #3498: when equations in algebraic loop?
Is it allowed for when equations to be part of an algebraic loop?
Following model leads to an error in Dymola but not in OMEdit:

```
model algebraicLoop 
	Real x;
	Real y(start=0.5,fixed=true);
	equation
		x = time - y;
		when time>0.5 then
		    y = -x/2;
		end when;
end algebraicLoop;
```
I have not found any statement to this regard in the specification. Please clarify.

### LINKED ISSUE #3499: when initial() with impure function?
E.g. the MSL sample Modelica.Blocks.Examples.Noise.ImpureGenerator uses Modelica.Blocks.Examples.Noise.Utilities.ImpureRandom which contains:

```
when {initial(), sample(samplePeriod,samplePeriod)} then
   y = Modelica.Math.Random.Utilities.impureRandom(globalSeed.id_impure);
end when;
```
During initialization the when initial() equations are active, so the result may depend on the number of iterations needed to solve the initial system. Is this desired?

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #3499
Closes #3498

### CODE REVIEW

File: chapters/equations.tex [hunk 1]
```diff
@@ -793,6 +794,9 @@ \section{Initialization, initial equation, and initial algorithm}\label{initiali
 
 Before the start of the integration, it must be guaranteed that for all variables \lstinline!v!, \lstinline!v = pre(v)!.
 If this is not the case for some variables \lstinline!vi!, \lstinline!pre(vi) := vi! must be set and an event iteration at the initial time must follow, so the model is re-evaluated, until this condition is fulfilled.
+In details this means that during initialization initial equations and normal equations are solved with \lstinline!v! and \lstinline!pre(v)! as unknowns without any event iterations.
```
Comment (henrikt-ma) [inline chapters/equations.tex:797]: Singular sounds better:
```suggestion
In detail this means that during initialization initial equations and normal equations are solved with \lstinline!v! and \lstinline!pre(v)! as unknowns without any event iterations.
```

File: chapters/equations.tex [hunk 2]
```diff
@@ -793,6 +794,9 @@ \section{Initialization, initial equation, and initial algorithm}\label{initiali
...
+Then \lstinline!v! is copied to \lstinline!pre(v)! and only the normal equations are solved repeatedly until \lstinline!v = pre(v)!.
+Tools may optimize this to not compute unnecessary \lstinline!pre(v)! and only perform the event iteration if necessary.
```
Comment (henrikt-ma) [inline chapters/equations.tex:799]: What is the purpose of writing out this possibility of optimization normatively?

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
In addition to other suggestions, I'd like to get rid of this non-normative text, now that we have clearly defined that it is not allowed:
```
\begin{nonnormative}
Comment: The semantics are undefined if the function call of an
impure function is part of an algebraic loop.
\end{nonnormative}
```

**Comment by HansOlsson** [comment]
Phone meeting:
Hans: 

If we want to have names:
1. Initialization, with additional variables (pre(v) as unknown and x for states) and additional equations (initial equations etc).
2. (Initial) Event iteration; iteration procedure at initial event
3. Start of continuous integration 
But probably no need to add.

File: chapters/equations.tex (hunk 2 above)
Comment (HansOlsson) [reply chapters/equations.tex:799]: Well, given that some only want to standardize the result before starting the continuous integration I think it is good to have this.

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
The request to remove obsolete text in https://github.com/modelica/ModelicaSpecification/pull/3532#pullrequestreview-2135684533 is still not addressed.  Apart from that, it looks OK.

File: chapters/equations.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/equations.tex:799]: OK, while I can understand the attempt to address these parties, it still looks odd to me to say such a thing normatively.  Hence:
```suggestion

\begin{nonnormative}
Tools may optimize this by not computing unnecessary \lstinline!pre(v)!, and only perform the event iteration if necessary.
\end{nonnormative}
```

**Review by henrikt-ma — DISMISSED** [review]
Same as last time:
> The request to remove obsolete text in https://github.com/modelica/ModelicaSpecification/pull/3532#pullrequestreview-2135684533 is still not addressed.  Apart from that, it looks OK.

If there is a reason why this text shouldn't be considered obsolete, please explain.

**Comment by HansOlsson** [comment]
> Same as last time:
> 
> > The request to remove obsolete text in [#3532 (review)](https://github.com/modelica/ModelicaSpecification/pull/3532#pullrequestreview-2135684533) is still not addressed.  Apart from that, it looks OK.
> 
> If there is a reason why this text shouldn't be considered obsolete, please explain.

There are several reasons to keep that:
1. It wasn't part of PR - this constant feature creep of blocking PR for parts that aren't in the PR needs to stop.
2. Since there are ways of circumventing the purity-check it is possible to call an impure function in a system of equations without it being an error.

**Comment by HansOlsson** [comment]
Pinging for reviews after holidays.

**Comment by henrikt-ma** [comment]
> * It wasn't part of PR - this constant scope creep of blocking PR for parts that aren't in the PR needs to stop.

It is related to this PR, as this PR is introducing this rule that looks crystal clear to me:
```
It is an error if an impure function call is part of a systems of equations (including linear systems), even if called in agreement with the restrictions above.
```

**Comment by HansOlsson** [comment]
> > * It wasn't part of PR - this constant scope creep of blocking PR for parts that aren't in the PR needs to stop.
> 
> It is related to this PR, as this PR is introducing this rule that looks crystal clear to me:
> 
> > It is an error if an impure function call is part of a systems of equations (including linear systems), even if called in agreement with the restrictions above.

I have opened a different PR #3571 to clarify that, since **as previously stated** it is an unrelated issue.

File: chapters/functions.tex [hunk 3]
```diff
@@ -280,9 +280,9 @@ \section{Pure Modelica Functions}\label{pure-modelica-functions}
   Binding equations for external objects.
 \end{itemize}
 
-For initial equations, initial algorithms, and bindings it is an error
-if the function calls are part of systems of equations and thus have to
-be called multiple times.
+It is an error if an impure function call is part of a systems of equations (including linear systems), even if called in agreement with the restrictions above.
+The reason is that solving systems of equations generally require expressions to be evaluated an unknown number of times.
```
Comment (eshmoylova) [inline chapters/functions.tex:284]: Typo: "solving ... requires"
```suggestion
The reason is that solving systems of equations generally requires expressions to be evaluated an unknown number of times.
```

**Review by eshmoylova — COMMENTED** [review]
Checking with the initialization team for any feedback.

**Review by eshmoylova — CHANGES_REQUESTED** [review]
Changes look good overall. Just requesting a typo fix.

**Review by eshmoylova — APPROVED** [review]

**Pull request merged**